### PR TITLE
Make Float16 available for macOS on Apple Silicon

### PR DESCRIPTION
**Pull request by stephentyrone**
Due to an unstable (and undesirable) calling convention in the LLVM layer for x86, I had previously marked Float16 unconditionally unavailable on macOS. My hope was that Intel would stabilize the calling convention and we could make it available on both macOS platforms at the same time. Unfortunately, that hasn't happened, and we want to make the type available for macOS/arm users.

So, I am making the availability mirror Float80--the type will be unavailable for macOS on x86_64, and available on all other platforms (the other x86 platforms don't have a binary-stability guarantee to worry about). This isn't ideal. In particular, if/when the calling conventions for Float16 stabilize in LLVM, we would want to make the type available, but it would then have _different_ availability for different architectures of macOS, which the current availability system is not well-equipped to handle (it's possible, but not very ergonomic). Nonetheless, this seems like the best option.

The good news is that because the full API is already built in Swift (and simply marked unavailable), we can simply add macOS 11.0 availability for these API and it will work.

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci please test

File: stdlib/public/core/FloatingPointParsing.swift.gyb [hunk 1]
```diff
@@ -136,7 +135,7 @@ extension ${Self}: LosslessStringConvertible {
   %if bits == 16:
     self.init(Substring(text))
   %else:
-    if #available(macOS 10.16, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+    if #available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
```
Comment (stephentyrone) [inline stdlib/public/core/FloatingPointParsing.swift.gyb:138]: @tbkka I think we made a change here that hasn't been pushed upstream for whatever reason--do you want to bring the two in sync now?

**Review by lorentey — APPROVED** [review]
Looks good! 👍 

Using these will be a pain on macOS for a while, but partial availability is still better than none at all.

File: stdlib/public/core/FloatingPointParsing.swift.gyb (hunk 1 above)
Comment (tbkka) [reply stdlib/public/core/FloatingPointParsing.swift.gyb:138]: Yes, we should sync these up.

Comment (stephentyrone) [reply stdlib/public/core/FloatingPointParsing.swift.gyb:138]: Do you want to put up a PR and tag me to review?

**Comment by xedin** [comment]
@stephentyrone @lorentey Looks like this has caused a regression in Surge and Kingfisher projects. Source compat suite failed here - https://github.com/apple/swift/pull/34834. Looks like main branch source compat suite just picked that commit up and going to start failing as well - https://ci.swift.org/view/Source%20Compatibility/job/swift-main-source-compat-suite/5600/

**Comment by xedin** [comment]
I'm not sure what the action plan is here - XFAIL the projects or fix `Accelerate`?

**Comment by lorentey** [comment]
Interesting! ci.swift.org currently builds with Xcode: 12.2 Beta 3 (12B5035g), whose Accelerate overlay includes the following:

```swift
@available(iOS 14, tvOS 14, watchOS 7, *)
@available(OSX, unavailable)
@available(macCatalyst, unavailable)
extension Float16 : Accelerate.BNNSScalar {
  public static var bnnsDataType: Accelerate.BNNSDataType {
    get
  }
}
```

This looks just fine to me. Perhaps the two flavors of unavailability differ somehow?

**Comment by lorentey** [comment]
Oh, I guess it could be because the x86_64 macOS interface now declares `Float16` unavailable everywhere, including non-macOS platforms.

**Comment by AnthonyLatsis** [comment]
> Oh, I guess it could be because the x86_64 macOS interface now declares Float16 unavailable everywhere, including non-macOS platforms.

Yeah. I think we should keep the old availability for the macOS x86_64 interface:
```swift
...

% if bits == 16:
@available(*, unavailable, message: "${Self} is not available on target platform.")
% else:
@available(iOS 14, tvOS 14, watchOS 7, *)
@available(OSX, unavailable)
@available(macCatalyst, unavailable)
% end
public struct ${Self} {

```

**Comment by philipturner** [comment]
@stephentyrone 

I'm trying to bring a GPU framework to macOS, where half precision is used in shaders. I can avoid doing actual `Float16` computations on the CPU, but I want to use `Float16` in the generic signature of a GPU-backed buffer. Here's my dilemma:

```swift
#if !((os(macOS) || targetEnvironment(macCatalyst)) && arch(x86_64))
test(Tensor<Float16>.incremented, input: 42, expected: 43)
#endif
test(Tensor<Float>.incremented, input: 42, expected: 43)
test(Tensor<Int8>.incremented, input: 42, expected: 43)
test(Tensor<Int16>.incremented, input: 42, expected: 43)
test(Tensor<Int32>.incremented, input: 42, expected: 43)
test(Tensor<UInt8>.incremented, input: 42, expected: 43)
test(Tensor<UInt16>.incremented, input: 42, expected: 43)
```

I want people with Intel Macs to take full advantage of half-precision processing power on their GPUs. I'm planning to create a [swift-intel-float16](https://github.com/philipturner/swift-intel-float16) repo, which brings `Float16` to Intel Macs through [conditional package dependencies](https://github.com/s4tf/s4tf/blob/c0ade64b60d7ddcfb1388fadc9e4cd9cadc1a223/Package.swift#L21-L37). It will play out similarly to [swift-reflection-mirror](https://github.com/philipturner/swift-reflection-mirror), which was Stdlib copypasta that enabled `@_spi(Reflection)` on release toolchains.

Is there anything I should look out for? I'm not trying to create ABI-stable Xcode frameworks. I just want users to compile my code via SwiftPM and play around with Swift for TensorFlow.

**Comment by stephentyrone** [comment]
If you don’t need ABI stability, you shouldn’t have any major issues.

**Pull request merged**